tests/ge_test.py

```diff
@@ -0,0 +1,247 @@
+'''
+Provides a Geo Engine instance for unit testing purposes.
+'''
+
+from collections.abc import Iterator
+from contextlib import contextmanager
+from pathlib import Path
+import random
+import string
+import subprocess
+import os
+import shutil
+import socket
+import threading
+from typing import Optional
+from dotenv import load_dotenv
+import psycopg
+
+TEST_CODE_PATH_VAR = 'GEOENGINE_TEST_CODE_PATH'
+
+POSTGRES_HOST = 'localhost'
+POSTGRES_PORT = 5432
+POSTGRES_DATABASE = 'geoengine'
+POSTGRES_USER = 'geoengine'
+POSTGRES_PASSWORD = 'geoengine'
+
+
+@contextmanager
+def GeoEngineTestInstance(port: Optional[int] = None) -> Iterator['GeoEngineProcess']:  # pylint: disable=invalid-name
+    '''Provides a Geo Engine instance for unit testing purposes.'''
+
+    load_dotenv()
+
+    if TEST_CODE_PATH_VAR not in os.environ:
+        raise RuntimeError(f'Environment variable {TEST_CODE_PATH_VAR} not set')
+
+    geo_engine_binaries = GeoEngineBinaries(Path(os.environ[TEST_CODE_PATH_VAR]))
+
+    try:
+        ge = GeoEngineProcess(
+            geo_engine_binaries=geo_engine_binaries,
+            port=get_open_port() if port is None else port,
+            db_schema=generate_test_schema_name(),
+        )
+        ge._start()  # pylint: disable=protected-access
+        yield ge
+    finally:
+        ge._stop()  # pylint: disable=protected-access
+
+
+def get_open_port() -> int:
+    '''Get an open port on the local machine.'''
+    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+        s.bind(("", 0))
+        return s.getsockname()[1]
+
+
+def generate_test_schema_name() -> str:
+    '''Generate a test schema name.'''
+    schema_name = 'pytest_'
+    for _ in range(10):
+        schema_name += random.choice(string.ascii_letters)
+    return schema_name
+
+
+class GeoEngineBinaries:
+    '''
+    Geo Engine binaries with `cargo` for testing.
+
+    This class is a singleton.
+    It builds the Geo Engine binaries from the given code path.
+    '''
+
+    _instance: Optional['GeoEngineBinaries'] = None
+    _lock = threading.Lock()
+
+    _code_path: Path
+    _server_binary_path: Path
+    _cli_binary_path: Path
+
+    def __new__(cls, code_path: Path) -> 'GeoEngineBinaries':
+        '''Create Geo Engine binaries for testing.'''
+
+        if cls._instance is not None:
+            return cls._instance
+
+        with cls._lock:
+            if cls._instance is None:
+                cls._instance = super().__new__(cls)
+                cls._instance._code_path = code_path
+
+                cls._instance._build_geo_engine()
+
+        return cls._instance
+
+    def _build_geo_engine(self) -> None:
+        '''Build the Geo Engine binaries.'''
+
+        cargo_bin = shutil.which('cargo')
+
+        if cargo_bin is None:
+            raise RuntimeError('Cargo not found')
+
+        subprocess.run(
```

Comment:
maybe log to console that this can take several minutes?

Reply:
[9fe1c44](https://github.com/geo-engine/geoengine-python/pull/211/commits/9fe1c4400a94f4fb04d15bb775bd4f246c11af5a)

Reply:
doesn't show up for me when I run pytest -v

Reply:
[1a1c96d](https://github.com/geo-engine/geoengine-python/pull/211/commits/1a1c96db2eb67445a584476225f9dd84815efa37)